s3torchbenchmarking/src/s3torchbenchmarking/dcp/benchmark.py

```diff
@@ -0,0 +1,154 @@
+        start_time = perf_counter()
+        multiprocessing.spawn(
+            run,
+            (
+                cfg.backend,
+                cfg.epochs,
+                benchmark_model.model,
+                world_size,
+                storage_writer,
+                save_timestamps,
+            ),
+            nprocs=world_size,
+            join=True,
+        )
+        end_time = perf_counter()
+        processing_times.append(end_time - start_time)
+
+        collector: List[float] = []
+        while not save_timestamps.empty():
+            collector.append(save_timestamps.get())
+        ptp_save_times.append(np.ptp(collector))
+
+    save_results(
+        cfg,
+        benchmark_model,
+        ptp_save_times=ptp_save_times,
+        processing_times=processing_times,
+        monitor=monitor,
+    )
+
+
+def validate_world_size(world_size: int) -> int:
+    """Enforce `world_size` to be within the current node's capacity.
+
+    FIXME: only works when the backend is "nccl"; what about "gloo" and CPUs?
+    """
+    device_count = torch.cuda.device_count()
+    if world_size > device_count:
+        logger.warning(
+            "Received a `world_size` of %i, while only %i are available: decreasing to %i",
+            world_size,
+            device_count,
+            device_count,
+        )
+        return device_count
+    return world_size
+
+
+def get_writer(cfg: DictConfig, suffix: str) -> FileSystemWriter:
+    """Instantiate a checkpoint writer based on the input config."""
+    if cfg.checkpoint.storage == "disk":
+        local_path = Path(cfg.path) / suffix
+        logger.info("Saving checkpoint to %s (local disk)...", local_path)
+        return dcp.FileSystemWriter(local_path, thread_count=cfg.thread_count)
+    elif cfg.checkpoint.storage == "s3":
+        uri = build_checkpoint_uri(cfg.s3.uri, suffix)
+        logger.info("Saving checkpoint to %s (S3)...", uri)
+        return S3StorageWriter(cfg.s3.region, uri, thread_count=cfg.thread_count)
+    raise ValueError(f"Storage writer {cfg.checkpoint.storage} not supported")
+
+
+def build_checkpoint_uri(s3_uri: str, suffix: str) -> str:
+    suffix = suffix.lstrip("/")
+    return s3_uri + suffix if s3_uri.endswith("/") else s3_uri + "/" + suffix
+
+
+def setup(backend: str, world_size: int, rank: int) -> None:
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12355"
+    dist.init_process_group(backend=backend, world_size=world_size, rank=rank)
+
+
+def run(
+    rank: int,  # needs to be passed first (provided by `multiprocessing.spawn` automatically)
+    backend: str,
+    epochs: int,
+    model: Module,
+    world_size: int,
+    storage_writer: FileSystemWriter,
+    save_timestamps: Queue,
+) -> None:
+    """Execute the actual code for checkpoint saving.
+
+    This function is meant to be executed in subprocesses."""
+
+    # FIXME: configure logging in subprocess accordingly
+
+    torch.cuda.set_device(rank)
```

Comment:
For world size greater than gpus this throws error.  Instead we should set this as follows
```
rank = dist.get_rank()
    device_id = rank % torch.cuda.device_count()
    torch.cuda.set_device(device_id)
```

Reply:
I tried that and got the following error:

```
torch.distributed.DistBackendError: NCCL error in: /opt/conda/conda-bld/pytorch_1716905969073/work/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1970, invalid usage (run with NCCL_DEBUG=WARN for details), NCCL version 2.20.5
ncclInvalidUsage: This usually reflects invalid usage of NCCL library.
Last error:
Duplicate GPU detected : rank 4 and rank 0 both on CUDA device 1b0
```

(with `world_size` of 8 on g5.12xlarge)
😢